x-pack/plugins/lens/public/app_plugin/shared/edit_on_the_fly/lens_configuration_flyout.tsx

```diff
@@ -256,7 +315,33 @@ export function LensEditConfigurationFlyout({
     hideLayerHeader: datasourceId === 'textBased',
     indexPatternService,
     setIsInlineFlyoutFooterVisible,
+    getUserMessages: getUserMessages(
+      datasourceMap[datasourceId],
+      datasourceStates[datasourceId].state,
+      activeVisualization,
+      visualization.state,
+      frameDatasourceAPI
+    ),
   };
+
+  // building expression can be expensive, that is why it's a function to run only if attributes changed and has no errors
+  const getNewStateExpression = () =>
+    visualization.state && visualization.activeId
+      ? buildExpression({
+          visualization: activeVisualization,
+          visualizationState: visualization.state,
+          datasourceMap,
+          datasourceStates,
+          datasourceLayers: frameDatasourceAPI.datasourceLayers,
+          indexPatterns: frameDatasourceAPI.dataViews.indexPatterns,
+          dateRange: frameDatasourceAPI.dateRange,
+          nowInstant: startDependencies.data.nowProvider.get(),
+          searchSessionId,
+        })
+      : null;
+
+  const isSaveable = attributesChanged && !!getNewStateExpression();
+
```

Comment:
My understanding is that it would be difficult to do because the other `buildExpression` call happens inside the embeddable :( I wouldn't like to update it from here because then embeddable would have to always run this side effect (to update the redux store or wherever we want to have this data) and I think it's a bad approach. I'll try to think about solving it different way, but if you have something else in mind, please let me know!

Reply:
No nothing in mind! I am just thinking if we can be more performant here. It is ok in my mind to have this bug in 8.11 if we need more time to think about it more. It is on tech preview and the users can always edit back and correct it.

Reply:
@stratoula I checked the performance of buildExpression function and it is in the end, very quick. I wouldn't bother to optimize it here, as it doesn't affect rendering time, just builds AST from the state. 
<img width="215" alt="Screenshot 2023-11-02 at 15 16 09" src="https://github.com/elastic/kibana/assets/4283304/d14f5c4d-77eb-4085-96d4-0c8e613d52d6">

Reply:
Still not fan of this solution tbh :( I think we should reconsider our architecture. Running things twice is not a good approach. Have you thought about other paths we can take here?

Reply:
Ok, I understand but this is a deeper problem that has its roots in how the editor communicates with the embeddable. I don't think rewriting architecture should be a part of this task because it would be too big (and we already have a [task for it](https://github.com/elastic/kibana/issues/166522)). It is not the first place we run code twice  nor it is heavier performance-wise than other places we don't prioritise so I am not convinced to look for a solution for optimise this only. 

About the solution you're proposing:
> Can this isSaveable be updated inside the buildExpression in order to not run it twice?
We could update the store, but embeddable doesn't have an access to the store so we won't get any improvement this way or we would have to zip up the another callback there.
We could memoize the whole function buildExpression. Let's keep in mind though that the execution for a double layer xy chart takes 0.5ms, so I don't think it's worth memoization that would only prevent us from executing this function one more time.
This solution would make `buildExpression` function effectual and more complicated to understand because there would be side effects (unless we would return `isSaveable` along with expression and not run callback inside) so that really doesn't make sense to me. 

> Have you thought about other paths we can take here?
All of the other paths I can think of base on improving communication between embeddable and Lens editor (whether it's 
1. sharing Dashboard store and Lens store (not a bad idea btw, we shouldn't have multiple stores)), 
2. changing the params of embeddable to accept already run expression, 
3. or using observables to get the expression from embeddable inside the editor. 

All of them are too complex in my opinion to do them right now without proper planning and considering all the things we want to improve for embeddable.